convert.go

```diff
@@ -68,40 +74,83 @@ func InterfaceToVariable(input interface{}) (ast.Variable, error) {
 	var mapVal map[string]interface{}
 	if err := hilMapstructureWeakDecode(input, &mapVal); err == nil {
 		elements := make(map[string]ast.Variable)
+		elemType := ast.TypeInvalid
 		for i, element := range mapVal {
 			varElement, err := InterfaceToVariable(element)
 			if err != nil {
 				return ast.Variable{}, err
 			}
 			elements[i] = varElement
+			if elemType == ast.TypeInvalid {
+				elemType = varElement.Type
+			} else if elemType != varElement.Type {
+				return ast.Variable{}, fmt.Errorf(
+					"inconsistent map element types; previously %s but found %s at key %s",
+					elemType.Printable(), varElement.Type.Printable(), i,
+				)
+			}
+		}
+
+		// If the map is empty then we'll default to it being of type
+		// string, just so we have some sort of element type.
+		if elemType == ast.TypeInvalid {
+			elemType = ast.TypeString
 		}
 
 		return ast.Variable{
-			Type:  ast.TypeMap{ast.TypeAny},
+			Type:  ast.TypeMap{elemType},
 			Value: elements,
 		}, nil
 	}
 
 	var sliceVal []interface{}
 	if err := hilMapstructureWeakDecode(input, &sliceVal); err == nil {
 		elements := make([]ast.Variable, len(sliceVal))
+		elemType := ast.TypeInvalid
 		for i, element := range sliceVal {
 			varElement, err := InterfaceToVariable(element)
 			if err != nil {
 				return ast.Variable{}, err
 			}
 			elements[i] = varElement
+
+			if varElement.Type == ast.TypeUnknown {
+				// Ignore unknown values for the purposes of element type inference
+				continue
+			}
+
+			if elemType == ast.TypeInvalid {
+				elemType = varElement.Type
+			} else if elemType != varElement.Type {
```

Comment:
This also needs to handle `TypeUnknown` as a special case, so that computed values can pass through.